### ISSUE TITLE
Admin broken on master

### ISSUE THREAD

**Issue by mitsuhiko** [lektor/lektor#390]
I get this on master:

```
Uncaught (in promise) TypeError: Cannot read property 'props' of undefined
    at Object.isActive (http://localhost:5000/admin/static/gen/app.js:9182:27)
    at http://localhost:5000/admin/static/gen/app.js:9241:29
    at Array.map (native)
    at Object.render (http://localhost:5000/admin/static/gen/app.js:9232:44)
    at http://localhost:5000/admin/static/gen/vendor.js:16301:22
    at measureLifeCyclePerf (http://localhost:5000/admin/static/gen/vendor.js:15581:13)
    at ReactCompositeComponentWrapper._renderValidatedComponentWithoutOwnerOrContext (http://localhost:5000/admin/static/gen/vendor.js:16300:26)
    at ReactCompositeComponentWrapper._renderValidatedComponent (http://localhost:5000/admin/static/gen/vendor.js:16327:33)
    at ReactCompositeComponentWrapper.performInitialMount (http://localhost:5000/admin/static/gen/vendor.js:15867:31)
    at ReactCompositeComponentWrapper.mountComponent (http://localhost:5000/admin/static/gen/vendor.js:15763:22)
```

Oddly enough the generated code seems completely wrong:

```
	  isActive: function isActive(field) {
	    var value = undefined.props.value;
	    if (value == null) {
	      value = undefined.props.placeholder;
	      if (value == null) {
	        return false;
	      }
	    }
```

Not sure why this is `undefined.props`; the source says `this.props`.

**Comment by mitsuhiko** [comment]
I downgraded to 2.3 temporarily where it works.

**Comment by nixjdm** [comment]
I haven't been able to reproduce this yet, unfortunately. I've tried in a few different environments (node versions, OSs, browsers). Can you give me any more details? Does this completely kill admin?

**Comment by tariquesani** [comment]
FWIW - I cannot reproduce it either. Versions of packages would be appreciated.

**Comment by mitsuhiko** [comment]
Reproduces if i just wipe node_modules and run `make build-js`. I can dump the versions installed later.

**Comment by mitsuhiko** [comment]
And yes, admin is completely unusable.

**Comment by nixjdm** [comment]
The line that's generated incorrectly specifically looks like that it's using ES5 for something where it should be using ES6. The fat arrow is ES6, but gives the error you have if it's read as if it's ES5. This particular fat arrow (and many others) was added in https://github.com/lektor/lektor/commit/33972ede#diff-43.

There are still some fat arrows and other features of ES6 used in 2.3 though, so I'm not sure why you don't get basically the same error when you use 2.3. I know there have been many changes in the package.json file.

I'm betting something is out of date or some version mismatch; node or some npm package. Perhaps we should try to control node versions with nvm, or put a note in the README.

**Comment by mitsuhiko** [comment]
Why does the node version matter here? Surely this goes through babel.

**Comment by nixjdm** [comment]
My understanding is that you can install packages with npm that rely on a particular version of node as an engine, even though you don't have the most compatible version of node installed. It's a big pain sometimes. So you can install package version x that is 100% compatible with node 6 (for instance), but only 80% compatible with node 4, yet you can still install it with node 4. One way this can happen is if the package relies on a certain NODE_MODULE_VERSION (node's api).

I'm not sure what the mismatch is here, but it looks to me like there probably is one. Issue #387 and issue #383 were fixed by just upgrading node, for instance. I'm not a fan of seeing this problem in node a lot.

**Comment by mitsuhiko** [comment]
But this is from the output from the babel build step. Not sure how this has anything to do with ES6.

**Comment by mitsuhiko** [comment]
FWIW, lektor should not require node newer than 6 or 7 or this is likely to cause a lot of issues for people.

**Comment by nixjdm** [comment]
I'm not sure how this error is happening or that this is an ES6 issue, just if it's a clue as to what's going on, you'd also get this error trying to use the fat arrow in ES5. It looks like babel is trying to transpile React with ES5 into regular ES5, where it should be transpiling React with ES6 into regular ES5.

**Comment by mitsuhiko** [comment]
So what happened is that babel was upgraded from 5 to 6 and in babel 6 the rules for fat arrow functions changed. In the widgets the code currently does `isActive: (field) => {` which is incorrect in babel 6 (should be a regular function instead).

So we either need to downgrade to babel 5 again or make the code work with babel 6.

I assume you do not see this behavior because for whatever reason you use babel 5?

**Comment by mitsuhiko** [comment]
Seems like @singingwolfboy introduced the bug in 33972ede849b5097d2fdb7b6c06925101d811c80

Some of these changes are definitely very wrong.

**Comment by nixjdm** [comment]
I may have reproduced this! I thought of it as I was falling asleep. I tried it and I get a different error that breaks all of admin, but if I understand it the rough cause is the same. For you to be getting what I've got you must be using npm 5.

npm 5 introduced the package-json.lock file, which freezes your dependency tree and is created and used automatically. If I have npm 5 installed, I checkout Lektor 2.3, and build-js, then I get a working admin and also a package-json.lock file. If I then checkout Lektor master and build-js, I get a broken admin. This is because npm is building the dependencies again from the package-json.lock file, and not from master's new package.json file. There are many changes in package.json that are necessary and if you didn't delete package-json.lock then they'll all be ignored.

You're right that stock Babel 6 won't transform arrows, and all of the default transforms were removed. This shouldn't be a problem though, because the es2015 preset was also added to the package.json file, and I believe the correct arrow plugins are wrapped up in that. I was using Babel 6 and got a working admin.

Try checking out master again and looking for that package-json.lock file. If you find it and it specifies Babel 5 and other things from 2.3 (or any older version) then removing that file and and doing a build-js again might fix your problem.

Crossing my fingers.

**Comment by mitsuhiko** [comment]
The package.json has an incorrect babel dependency (babel 6). It won't work with that one.

https://github.com/lektor/lektor/blob/7de4da2c37b1db73c12ec81c7c502533e28b2d23/lektor/admin/package.json#L7-L14

I do not use npm 5 here because it has a few issues that cause problems in other projects so there is no lock file that would fix this. There is also no lock file checked into the repository. Even if there were, the lock file would if it specifies babel 5 conflict with the babel 6 from the package.json.

**Comment by torian257x** [comment]
see here to reproduce error

https://github.com/Jossnaz/lektor_bug390

**Comment by nixjdm** [comment]
I think I understand this now. @Jossnaz's test repo made me realize that I wasn't seeing this error because I was using a test project that was too simple, the example project in this repo, IIRC. However, his test repo, and also the lektor-website repo make use of portions of the compiled js that have this kind of error in them. 

The fat arrow is indeed the problem, because it's use of `this` is different than a normal function, i.e., [an arrow function doesn't create it's own `this`, but uses its parent's](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions#No_binding_of_this).

I talked to @singingwolfboy and most of the changes in this commit were by running the linter with --fix, but he thinks he made some changes by hand too. Maybe these were it, or it was the linter. [The linter used to have some rules about arrows that I think have been totally removed now](https://standardjs.com/changelog.html). Either way I guess we shouldn't trust linters completely. I'm going through the additional arrows in https://github.com/lektor/lektor/commit/33972ede849b5097d2fdb7b6c06925101d811c80 and checking them individually right now. Most (all?) of them need to be changed back to normal functions. There are good uses of arrow functions in the codebase, but most (all?) of these particular ones are bad. It hinges on their use of `this`. I think I'll have a pull request up this weekend finally.

**Comment by nixjdm** [comment]
I have a [pull request](https://github.com/lektor/lektor/pull/427) up for this now. I did check all the arrow functions added by 33972ed. Most were actually fine. Most that used `this` were not though, and they were reverted.

**Comment by nixjdm** [comment]
Fixed in #427
